sogs/model/filter.py

```diff
@@ -0,0 +1,212 @@
+from .exc import InvalidData
+
+from typing import Optional, List, Union
+import time
+import contextlib
+
+
+class SimpleFilter:
+    """
+    Class representing a simple word filter searching for naughty words
+
+    Object Properties:
+        bot - bot this filter is servicing
+        current_message - reference to current data being analyzed
+    """
+
+    def __init__(self, _bot: Bot, _room: Room = None):
+        self.bot: Bot = _bot
+        self.current_message: Post = None
+
+    def filtering(self):
+        settings = {
+            'profanity_filter': config.PROFANITY_FILTER,
+            'profanity_silent': config.PROFANITY_SILENT,
+            'alphabet_filters': config.ALPHABET_FILTERS,
+            'alphabet_silent': config.ALPHABET_SILENT,
+        }
+        if self.token in config.ROOM_OVERRIDES:
+            for k in (
+                'profanity_filter',
+                'profanity_silent',
+                'alphabet_filters',
+                'alphabet_silent',
+            ):
+                if k in config.ROOM_OVERRIDES[self.token]:
+                    settings[k] = config.ROOM_OVERRIDES[self.token][k]
+        return settings
+
+    def filter_should_reply(self, filter_type, filter_lang, room: Room):
+        """If the settings say we should reply to a filter, this returns a tuple of
+
+        (profile name, message format, whisper)
+
+        where profile name is the name we should use in the reply, message format is a string with
+        substitutions ready, and whisper is True/False depending on whether it should be whispered
+        to the user (True) or public (False).
+
+        If we shouldn't reply this returns (None, None, None)
+        """
+
+        if not config.FILTER_SETTINGS:
+            return (None, None, None)
+
+        reply_format = None
+        profile_name = 'SOGS'
+        public = False
+
+        # Precedences from least to most specific so that we load values from least specific first
+        # then overwrite them if we find a value in a more specific section
+        room_precedence = ('*', room.token)
+        filter_precedence = ('*', filter_type, filter_lang) if filter_lang else ('*', filter_type)
+
+        for r in room_precedence:
+            s1 = config.FILTER_SETTINGS.get(r)
+            if s1 is None:
+                continue
+            for f in filter_precedence:
+                settings = s1.get(f)
+                if settings is None:
+                    continue
+
+                rf = settings.get('reply')
+                pn = settings.get('profile_name')
+                pb = settings.get('public')
+                if rf is not None:
+                    reply_format = random.choice(rf)
+                if pn is not None:
+                    profile_name = pn
+                if pb is not None:
+                    public = pb
+
+        return (reply_format, profile_name, public)
+
+    def read_message(self, user: User, data: bytes, room: Room):
```

Comment:
naming nitpick: `view_message` or `inspect_message` feel a bit less ambiguous. (does the read the message or does it tell us if we read the message?)

Reply:
this class is holding all the previous filter functionalities that we can turn into a bot to test the interface